project-clone/internal/git/setup.go

```diff
@@ -18,46 +18,68 @@ package git
 import (
 	"fmt"
 	"log"
+	"os"
+	"path"
 
-	"github.com/devfile/api/v2/pkg/apis/workspaces/v1alpha2"
+	dw "github.com/devfile/api/v2/pkg/apis/workspaces/v1alpha2"
 
 	"github.com/devfile/devworkspace-operator/project-clone/internal"
 )
 
-func SetupGitProject(project v1alpha2.Project) error {
+func SetupGitProject(project dw.Project) error {
 	needClone, needRemotes, err := internal.CheckProjectState(&project)
 	if err != nil {
 		return fmt.Errorf("failed to check state of repo on disk: %s", err)
 	}
 	if needClone {
-		err := CloneProject(&project)
-		if err != nil {
-			return fmt.Errorf("failed to clone project: %s", err)
-		}
-		repo, err := internal.OpenRepo(&project)
-		if err != nil {
-			return fmt.Errorf("failed to open existing project in filesystem: %s", err)
-		} else if repo == nil {
-			return fmt.Errorf("unexpected error while setting up remotes for project: git repository not present")
-		}
-		if err := SetupRemotes(repo, &project); err != nil {
-			return fmt.Errorf("failed to set up remotes for project: %s", err)
-		}
-		if err := CheckoutReference(repo, &project); err != nil {
-			return fmt.Errorf("failed to checkout revision: %s", err)
-		}
+		return doInitialGitClone(&project)
 	} else if needRemotes {
-		repo, err := internal.OpenRepo(&project)
-		if err != nil {
-			return fmt.Errorf("failed to open existing project in filesystem: %s", err)
-		} else if repo == nil {
-			return fmt.Errorf("unexpected error while setting up remotes for project: git repository not present")
-		}
-		if err := SetupRemotes(repo, &project); err != nil {
-			return fmt.Errorf("failed to set up remotes for project: %s", err)
-		}
+		return setupRemotesForExistingProject(&project)
 	} else {
 		log.Printf("Project '%s' is already cloned and has all remotes configured", project.Name)
+		return nil
+	}
+}
+
+func doInitialGitClone(project *dw.Project) error {
+	// Clone into a temp dir and then move set up project to PROJECTS_ROOT to try and make clone atomic in case
+	// project-clone container is terminated
+	tmpClonePath := path.Join(internal.CloneTmpDir, internal.GetClonePath(project))
```

Comment:
how much seconds does it take to move from tmp to the cloned directory ?

Reply:
Depends on a variety of factors, but it's a linux move at the directory level so presumably almost no time -- it's just modifying the path.

Reply:
ok I see, temp directory is inside the PROJECTS_ROOT folder and not like in default temporary directory of the system

Reply:
Yeah, I had to keep it that way otherwise we're copying files over one-by-one and risk getting stuck in a half-setup repo as before. The new flow is
1. Create `$PROJECTS_ROOT/project-clone-<random-string>` directory to hold temporary files
2. Trap on exit to remove temp dir, so that if we get killed early no changes are made to the PVC
3. Clone/unzip to `$PROJECTS_ROOT/project-clone-<random-string>/project-name`
4. Move `project-name` to `$PROJECTS_ROOT`
5. Temp dir is removed on exit